### ISSUE TITLE
Add MAX_EXECUTION_TIME optimizer hints in MySQL

### ISSUE THREAD

**Issue by SeongJaeSong** [sequelize/sequelize#15173]
## Issue Creation Checklist

- [X] I understand that my issue will be automatically closed if I don't fill in the requested information
- [X] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

### Describe the feature you'd like to see implemented

Set max_execution_time per session when after connecting by adding `maxExecutionTime` option.

```js
new Sequelize(db, user, pw, {
  ...,
  timezone: '+09:00',
  maxExecutionTime: 60000,
  ...,
})
```

### Describe why you would like this feature to be added to Sequelize

For now, we should use raw query within `afterConnect` hook like below in order to set `max_execution_time` to every connection. It might be verbose.
```js
hooks: {
    afterConnect: async (connection: Connection, _config) => {
      await connection
        .promise()
        .query('SET SESSION max_execution_time = 60000;')
    },
  }
```

We can do this easier like `timezone` setting in Options for the constructor of Sequelize main class.

### Is this feature dialect-specific?

- [ ] No. This feature is relevant to Sequelize as a whole.
- [X] Yes. This feature only applies to the following dialect(s): mysql

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [X] Yes, I have the time but I will need guidance.
- [ ] No, I don't have the time, but my company or I are [supporting Sequelize through donations on OpenCollective](https://opencollective.com/sequelize).
- [ ] No, I don't have the time, and I understand that I will need to wait until someone from the community or maintainers is interested in implementing my feature.

---

*Indicate your interest in the addition of this feature by adding the 👍 reaction. Comments such as "+1" will be removed.*

**Comment by SeongJaeSong** [comment]
@fzn0x
Can I get some information or explanation about why this option(max_execution_time) usage is recommended at query level?


And I agree with "it's only low effort to set" but, isn't it verbose? Is there a reason why we are reluctant to include it as the global option in `Sequelize` constructor?

**Comment by fzn0x** [comment]
> @fzn0x Can I get some information or explanation about why this option(max_execution_time) usage is recommended at query level?
> 
> And I agree with "it's only low effort to set" but, isn't it verbose? Is there a reason why we are reluctant to include it as the global option in `Sequelize` constructor?

Setting max_execution_time globally as ephys said it will be impactful to other systems which using the same database, you may need to reset the state back again when it's only used for certain queries, the best option is to set the timeout in query level, which is only the query that's really need it, if it's allowed to be part of `Sequelize` constructor is the same as we allowed something may be dangerous.

**Comment by SeongJaeSong** [comment]
@ephys 
I think we need only `per session` option not `global`. And per session option will occur overhead each time when creating a new connection.

But, whether use it(when need `per session` option) or not is up to user(developer). So I think if we explain about the overhead in comments or docs, it doesn't matter.

**Comment by SeongJaeSong** [comment]
> Setting max_execution_time globally as ephys said it will be impactful to other systems which using the same database, you may need to reset the state back again when it's only used for certain queries, the best option is to set the timeout in query level, which is only the query that's really need it, if it's allowed to be part of `Sequelize` constructor is the same as we allowed something may be dangerous.

Yes, I agree with that(Set max_execution_time globally). And I suggest about Set max_execution_time per session that created by `Sequelize`. I think it doesn't make any side effect to other systems which using the same database.

**Comment by fzn0x** [comment]
I think that's would work

**Comment by ephys** [comment]
> Is there a reason why we are reluctant to include it as the global option in Sequelize constructor?

For a few reasons:
1. It's specific to MySQL, the option doesn't exist in other dialects
2. It's trivial to implement yourself through the afterConnect hook or by just
3. We'd have to choose between having it be global or per session, or let the user choose.

Because of all of these, I believe the cost to the ORM outweight the benefits. reason 3 is less important. I'm generally not for adding features that don't pass both reasons 1 & 2

Adding something like `mysqlMaxExecutionTimeHint` at the query level makes sense because it's currently impossible for users to add the hint without it

**Comment by SeongJaeSong** [comment]
@ephys 
Thank you for your detailed explanation.

And I have some more questions.
You said adding `mysqlMaxExecutionTimeHint` make sense because it is necessary to add the hint.
But max_execution_hint is specific to MySQL too, and we can add the hint with raw query(`sequelize.query(...)`) - I know this case is obviously more verbose than the feature I suggested.
I think this is against the reason 1 & 2. What do you think?

And how do we know the usage or demand for certain features?

**Comment by ephys** [comment]
No I said we can consider adding `mysqlMaxExecutionTimeHint` because that feature would be impossible to use if we don't provide support for it. Users can't modify the SQL that will be executed.

**Comment by SeongJaeSong** [comment]
I got it. Then, can we develop the feature `mysqlMaxExecutionTimeHint`?
